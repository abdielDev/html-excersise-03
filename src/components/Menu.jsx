import React, { useState, useEffect } from 'react';
import '../css/components/ButtonMenu.css'
import '../css/components/Menu.css'

const Menu = () => {
  const [count, setCount] = useState(0);
  const eventComposed = new Event('composed');
  const eventButtons = new Event('changeColorButtons');

  const handleClick = (e, eventComposed, eventButtons) => {
    const elementWithActive = document.querySelector('.Active');
    elementWithActive.classList.remove('Active')
    const { target } = e;
    e.target.dispatchEvent(eventComposed);
    e.target.dispatchEvent(eventButtons);
    target.classList.add('Active');

    setCount(count + 1);

    console.log(`Id: ${target.id}, Count: ${count}`);
  };

  let $lis;

  useEffect(() => {

    $lis = document.querySelectorAll('li.Button-Menu');

    $lis.forEach((node) => {
      node.addEventListener('composed', e => {
        e.stopPropagation();
        console.log('I\'m not propagating')
      });
      node.addEventListener('changeColorButtons', e => {
        const rgb1 = Math.random() * 256;
        const rgb2 = Math.random() * 256;
        const rgb3 = Math.random() * 256;
        let $anchors = document.querySelectorAll('button.Button');
        $anchors.forEach($anchor => {
          $anchor.style.backgroundColor = `rgb(${rgb1}, ${rgb2}, ${rgb3})`;
        })
      });
    });
  }, []);

  return (
    <ul className="Menu">
      <li onClick={e => handleClick(e, eventComposed, eventButtons)} id="home" className="Active Button-Menu">Inicio</li>
      <li onClick={e => handleClick(e, eventComposed, eventButtons)} id="about" className="Button-Menu">Acerca</li>
      <li onClick={e => handleClick(e, eventComposed, eventButtons)} id="services" className="Button-Menu">Servicios</li>
      <li onClick={e => handleClick(e, eventComposed, eventButtons)} id="products" className="Button-Menu">Productos</li>
      <li onClick={e => handleClick(e, eventComposed, eventButtons)} id="contact" className="Button-Menu">Contacto</li>
    </ul>
  );
};

export default Menu;