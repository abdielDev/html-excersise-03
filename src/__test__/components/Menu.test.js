import React from 'react';
import { shallow, mount } from 'enzyme';
import { create } from 'react-test-renderer';
import Menu from '../../components/Menu';

describe('<Menu />', () => {
  const menu = shallow(<Menu />);
  test('Render del componente Menu', () => {
    expect(menu.length).toEqual(1);
  });

  test('Render 5 listed elements in the menu', () => {
    expect(menu.find('ul').children()).toHaveLength(5);
  });
});

describe('Menu Behavior', () => {
  window.console.log = jest.fn();
  const menu = mount(<Menu />, { attachTo: document.body });
  const firstLi = menu.find('li').first()
  test('Oprimir botones del menu e imprimir count, id y \'I\'m not propagating', () => {
    firstLi.simulate('click');
    expect(window.console.log).toHaveBeenCalledTimes(2);
  });
  test('li element with class Active', () => {
    expect(menu.exists('.Active')).toBeTruthy();
  });
});

describe('Menu Snapshot', () => {
  test('Comprobar la UI del componente Menu', () => {
    const menu = create(<Menu />);
    expect(menu.toJSON()).toMatchSnapshot();
  });
});