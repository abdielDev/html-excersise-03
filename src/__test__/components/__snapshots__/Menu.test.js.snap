// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Menu Snapshot Comprobar la UI del componente Menu 1`] = `
<ul
  className="Menu"
>
  <li
    className="Active Button-Menu"
    id="home"
    onClick={[Function]}
  >
    Inicio
  </li>
  <li
    className="Button-Menu"
    id="about"
    onClick={[Function]}
  >
    Acerca
  </li>
  <li
    className="Button-Menu"
    id="services"
    onClick={[Function]}
  >
    Servicios
  </li>
  <li
    className="Button-Menu"
    id="products"
    onClick={[Function]}
  >
    Productos
  </li>
  <li
    className="Button-Menu"
    id="contact"
    onClick={[Function]}
  >
    Contacto
  </li>
</ul>
`;
